src/unix/notbsd/linux/other/mod.rs

```diff
@@ -1,5 +1,104 @@
 pub type __priority_which_t = ::c_uint;
 
+cfg_if! {
+    if #[cfg(libc_union)] {
+        #[repr(C)]
+        #[derive(Copy, Clone)]
+        struct addr_bnd_t {
+            lower: *mut ::c_void,
+            upper: *mut ::c_void,
+        }
+
+        #[repr(C)]
+        #[derive(Copy, Clone)]
+        union sigfault_t_anonymous_union {
+            addr_bnd: addr_bnd_t,
+            pkey: u32,
+        }
+
+        #[repr(C)]
+        #[derive(Copy, Clone)]
+        struct sigfault_t {
+            addr: *mut ::c_void,
+            #[cfg(target_arch = "sparc")]
+            trapno: ::c_int,
+            addr_lsb: ::c_short,
+            anonymous_union: sigfault_t_anonymous_union,
+        }
+
+        #[repr(C, packed)]
+        #[derive(Copy, Clone)]
+        union sifields_t {
+            _pad: [::c_int; 29],
+            sigfault: sigfault_t,
+        }
+
+        impl ::fmt::Debug for sigfault_t_anonymous_union {
+            fn fmt(&self, f: &mut ::fmt::Formatter) -> ::fmt::Result {
+                // TODO: no idea how to tell which of the members is used
+                f.debug_struct("sigfault_t_anonymous_union")
+                    .finish()
+            }
+        }
+
+        impl ::fmt::Debug for sigfault_t {
+            fn fmt(&self, f: &mut ::fmt::Formatter) -> ::fmt::Result {
+                // TODO: include trapno on Sparc
+                f.debug_struct("sigfault_t")
+                    .field("addr", &self.addr)
+                    .field("addr_lsb", &self.addr_lsb)
+                    .field("anonymous_union", &self.anonymous_union)
+                    .finish()
+            }
+        }
+
+        impl ::fmt::Debug for sifields_t {
+            fn fmt(&self, f: &mut ::fmt::Formatter) -> ::fmt::Result {
+                // No way to print anything more detailed without the
+                // discriminant from siginfo_t
+                f.debug_struct("sifields_t").finish()
+            }
+        }
+    } else {
+        type sifields_t = [::c_int; 29];
+    }
+}
+
+s_no_extra_traits! {
+    pub struct siginfo_t {
+        pub si_signo: ::c_int,
+        pub si_errno: ::c_int,
+        pub si_code: ::c_int,
+        sifields: sifields_t,
+
+        #[cfg(target_arch = "x86_64")]
+        _align: [u64; 0],
+        #[cfg(not(target_arch = "x86_64"))]
+        _align: [usize; 0],
+    }
+}
+
+cfg_if! {
+    if #[cfg(libc_union)] {
+        impl siginfo_t {
+            pub unsafe fn si_addr(&self) -> *mut ::c_void {
+                self.sifields.sigfault.addr
+            }
+        }
+    }
```

Comment:
This method should also be implemented if `#[cfg(libc_union)]` is false.

Reply:
How would you want it implemented? `panic!()`? Manually compute the offset and hardcode it?

Reply:
> How would you want it implemented?

Such that it always works correctly.

> Manually compute the offset and hardcode it?

All `repr(C)` `union` fields are at offset 0, or what offset do you mean?

Reply:
Sure, in this particular case it's easy, it's just the first `*mut c_void` in `_pad`, however if we get to exposing other fields the logic will be more complicated. In any event, this one is easy enough.